File: api/queries/likes.js
let { db } = require('../database');
let format = require('pg-format');


function createLike(req, res, next) {

  const now = new Date();

  let body = {
    post_id: req.body.post_id,
    user_id: req.user.id
  };


  if (!body.post_id) {

    res.status(400)
      .json({
        status: 'error',
        message: 'post_id can not be null.'
      });

  } else {

    db.one(format('INSERT into post_likes(user_id, post_id, created, updated) ' +
      'values(%1$L, %2$L, %3$L, %3$L) ' +
      'RETURNING post_likes.id AS like_id',
      body.user_id, body.post_id, now))
      .then((response) => {

        res.status(201)
          .json({

            status: 'success',
            like_id: response.like_id,
            message: 'Inserted one like from user ' + body.user_id + ' to post ' + body.post_id

          });

      })
      .catch(function (err) {

        return next(err);

      });

  }

}

function deleteLike(req, res, next) {

  const like_id = parseInt(req.params.id);

  findLikeById(like_id, (err, like) => {

    if (err) {

      next(err);

    } else if (!req.user || parseInt(like.user_id) !== req.user.id) {

      res.status(403)
        .json({
          status: 'error',
          message: 'You are not authorized to delete this like.'
        });

    } else {

      db.result(format('DELETE FROM post_likes WHERE id = %L', like_id))
        .then(function (result) {

          /* jshint ignore:start */
          res.status(200)
            .json({

              status: 'success',
              message: `Removed like ${like_id}`

            });
          /* jshint ignore:end */

        })
        .catch(function (err) {

          return next(err);

        });

    }

  });

}

function findLikeById(id, cb) {

  db.oneOrNone(format('SELECT * FROM post_likes WHERE id = %L', id))

    .then((data) => {

      return cb(null, data);

    })
    .catch((err) => {

      return cb(err, null);

    })

}

function getLikes(req, res, next) {

  if (req.query.post_id) {

    const post_id = parseInt(req.query.post_id);

    db.any(format('SELECT post_likes.id, post_likes.user_id, post_likes.post_id, post_likes.created, post_likes.updated, ' +
      'users.login, users.firstname, users.lastname ' +
      'FROM post_likes INNER JOIN users ON post_likes.user_id = users.id ' +
      'WHERE post_likes.post_id = %L',
      post_id))

      .then((data) => {

        if (Array.isArray(data)) {

          return (serializeLikes(data));

        } else {

          return (serializeLike(data));

        }

      })
      .then((data) => {

        res.status(200)
          .json({

            status: 'success',
            data,
            message: 'Retrieved likes'

          });

      })
      .catch(function (err) {

        return next(err);

      });

  } else {

    res.status(400)
      .json({

        status: 'error',
        message: 'post_id not provided.'

      });

  }

}

function serializeLikes(data) {

  let newData = [];

  for (let i = 0, len = data.length; i < len; i++) {

    newData.push(serializeLike(data[i]));

  }

  return (newData);

}

function serializeLike(data) {

  return ({
    id: data.id,
    post_id: data.post_id,
    created: data.created,
    updated: data.updated,
    user: {
      id: data.user_id,
      login: data.login,
      firstname: data.firstname,
      lastname: data.lastname
    }
  });

}

// Add query functions
module.exports = {
  createLike,
  deleteLike,
  findLikeById,
  getLikes
};

File: go-api/queries/likes.go
package queries

import (
	"github.com/labstack/echo"
	"time"
	"../types"
	"net/http"
	"../db"
	"fmt"
	"strconv"
	"github.com/jackc/pgx"
	"github.com/jackc/pgx/pgtype"
)

func CreateLike(c echo.Context) error {

	now := time.Now()

	var body types.LikeForm; var err error
	c.Bind(&body)
	loggedId := getUserId(c)

	if body.PostId <= 0 {

		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "post_id can not be null.",
		})

	} else {

		err = db.App.Db.QueryRow("INSERT into post_likes(user_id, post_id, created, updated) "+
			"VALUES($1, $2, $3, $3) "+
			"RETURNING post_likes.id AS like_id",
			loggedId, body.PostId, now).Scan(&body.Id)

		if err != nil || body.Id <= 0 {

			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":  "error",
				"message": err.Error(),
			})
		}

		return c.JSON(http.StatusCreated, echo.Map{
			"status":  "success",
			"like_id": body.Id,
			"message": fmt.Sprintf("Inserted one like from user %v to post %v", loggedId, body.PostId),
		})
	}
}

func DeleteLike(c echo.Context) error {

	likeId, ok := strconv.ParseInt(c.Param("id"), 10, 64)
	loggedId := getUserId(c)

	if ok != nil {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":   "error",
			"message":  "Id not provided",
		})
	}

	like, err := findLikeById(likeId)

	if err != nil {
		if err != pgx.ErrNoRows {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":   "error",
				"message":  err.Error(),
			})
		} else {
			return c.JSON(http.StatusNotFound, echo.Map{
				"status":   "error",
				"message":  "Like not found",
			})
		}
	} else if like.UserId.Status != pgtype.Present || int64(like.UserId.Int) != loggedId {
		return c.JSON(http.StatusForbidden, echo.Map{
			"status":   "error",
			"message":  "You can't delete this like",
		})
	} else {

		_, err := db.App.Db.Exec("DELETE FROM post_likes WHERE id = $1", like.Id.Int)

		if err != nil {

			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":  "error",
				"message": err.Error(),
			})

		} else {
			message := fmt.Sprintf("Removed like %v", like.Id.Int)

			return c.JSON(http.StatusOK, echo.Map{
				"status":  "success",
				"message": message,
			})
		}
	}
}

func findLikeById(id int64) (types.Like, error) {

	var like types.Like

	err := db.App.Db.QueryRow("SELECT * FROM post_likes WHERE id = $1", id).
		Scan(&like.Id, &like.UserId, &like.PostId, &like.Created, &like.Updated)

	if err != nil {
		if err == pgx.ErrNoRows {
			return types.Like{}, err
		} else {
			CheckErr(err)
			return types.Like{}, err
		}
	}

	return like, nil
}

func GetLikes(c echo.Context) error {

	var postId int64; var ok, err error; var results *pgx.Rows; var likes []types.SingleLikeDataResponse
	postId, ok = strconv.ParseInt(c.QueryParam("post_id"), 10, 64)

	if ok != nil  {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "post_id not provided.",
		})
	} else {
		results, err = db.App.Db.Query("SELECT post_likes.id, post_likes.user_id, post_likes.post_id, post_likes.created, post_likes.updated, " +
			"users.login, users.firstname, users.lastname " +
			"FROM post_likes INNER JOIN users ON post_likes.user_id = users.id " +
			"WHERE post_likes.post_id = $1", postId)
	}

	if err != nil && err != pgx.ErrNoRows {
		return c.JSON(http.StatusInternalServerError, echo.Map{
			"status":   "error",
			"message":  err.Error(),
		})
	}

	for results.Next() {
		var like types.SingleLikeData
		if err = results.
			Scan(&like.Id, &like.UserId, &like.PostId, &like.Created, &like.Updated,
			&like.User.Login, &like.User.Firstname, &like.User.Lastname); err != nil {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":   "error",
				"message":  err.Error(),
			})
		}

		var singleLike types.SingleLikeDataResponse

		if like.Id.Status == pgtype.Present {
			singleLike.Id = int64(like.Id.Int)
		}
		if like.UserId.Status == pgtype.Present {
			singleLike.UserId = int64(like.UserId.Int)
			singleLike.User.Id = int64(like.UserId.Int)
		}
		if like.PostId.Status == pgtype.Present {
			singleLike.PostId = int64(like.PostId.Int)
		}
		if like.Created.Status == pgtype.Present {
			singleLike.Created = &(like.Created.Time)
		}
		if like.Updated.Status == pgtype.Present {
			singleLike.Updated = &(like.Updated.Time)
		}
		if like.User.Login.Status == pgtype.Present {
			singleLike.User.Login = like.User.Login.String
		}
		if like.User.Firstname.Status == pgtype.Present {
			singleLike.User.Firstname = like.User.Firstname.String
		}
		if like.User.Lastname.Status == pgtype.Present {
			singleLike.User.Lastname = like.User.Lastname.String
		}

		likes = append(likes, singleLike)
	}
	if err = results.Err(); err != nil {
		return c.JSON(http.StatusInternalServerError, echo.Map{
			"status":   "error",
			"message":  err.Error(),
		})
	}

	return c.JSON(http.StatusOK, echo.Map{
		"status":   "success",
		"data":     likes,
		"message":  "Retrieved likes",
	})
}
